zephyr/wrapper.c

```diff
@@ -326,14 +326,32 @@ int poll_for_register_delay(uint32_t reg, uint32_t mask,
 	return 0;
 }
 
+/* Mutable, volatile and not static to escape optimizers and static
+ * analyzers.
+ */
+volatile int *_sof_fatal_null = NULL;
+
 void k_sys_fatal_error_handler(unsigned int reason,
 			       const z_arch_esf_t *esf)
 {
 	ARG_UNUSED(esf);
+
+	/* flush and switch to immediate mode */
 	LOG_PANIC();
 
 	ipc_send_panic_notification();
 
+#if defined(CONFIG_ARCH_POSIX) || defined(CONFIG_ZEPHYR_POSIX)
+	LOG_ERR("Halting emulation");
+
+	/* In emulation we want to stop _immediately_ and print a useful stack
+	 * trace; not a useless pointer to some signal handler or Zephyr
+	 * cleanup routine. See Zephyr POSIX limitations discussed in:
+	 * https://github.com/zephyrproject-rtos/zephyr/pull/68494
+	 */
+	*_sof_fatal_null = 42;
```

Comment:
I usually do it with something as simple as `*(int *)NULL = 0;` Any reason this wouldn't be enough? Never failed me until now, except on some particular hardware platforms like #6093

Reply:
Optimizers and modern code flow analysis tend to make a mess of code that does undefined things.  Putting it in an extern guarantees (at least until LTO lands) that the compiler can't "fix" this.  The fundamental problem is that the desired behavior here isn't in fact portable: there's no way under the C standard to "crash in a way that ASAN detects before the OS".